pkg/sql/catalog/hydratedtables/hydratedcache.go

```diff
@@ -0,0 +1,327 @@
+var CacheSize = settings.RegisterNonNegativeIntSetting(
+	"sql.catalog.hydrated_tables.cache_size",
+	"number of table descriptor versions retained in the hydratedtables LRU cache",
+	128)
+
+// NewCache constructs a new Cache.
+func NewCache(settings *cluster.Settings) *Cache {
+	c := &Cache{
+		settings: settings,
+		metrics:  makeMetrics(),
+	}
+	c.mu.cache = cache.NewOrderedCache(cache.Config{
+		Policy: cache.CacheLRU,
+		ShouldEvict: func(size int, key, value interface{}) bool {
+			return size > int(CacheSize.Get(&settings.SV))
+		},
+		OnEvicted: func(key, value interface{}) {
+			putCacheKey(key.(*cacheKey))
+		},
+	})
+	return c
+}
+
+type hydratedTableDescriptor struct {
+	tableDesc *tabledesc.Immutable
+	typeDescs []*cachedType
+}
+
+type cachedType struct {
+	name    tree.TypeName
+	typDesc catalog.TypeDescriptor
+}
+
+type cacheKey lease.IDVersion
+
+func (c cacheKey) Compare(comparable llrb.Comparable) int {
+	o := comparable.(*cacheKey)
+	switch {
+	case c.ID < o.ID:
+		return -1
+	case c.ID > o.ID:
+		return 1
+	default:
+		switch {
+		case c.Version < o.Version:
+			return -1
+		case c.Version > o.Version:
+			return 1
+		default:
+			return 0
+		}
+	}
+}
+
+var _ llrb.Comparable = (*cacheKey)(nil)
+
+var cacheKeySyncPool = sync.Pool{
+	New: func() interface{} { return new(cacheKey) },
+}
+
+func newCacheKey(table catalog.TableDescriptor) *cacheKey {
+	k := cacheKeySyncPool.Get().(*cacheKey)
+	*k = cacheKey{
+		ID:      table.GetID(),
+		Version: table.GetVersion(),
+	}
+	return k
+}
+
+func putCacheKey(k *cacheKey) {
+	*k = cacheKey{}
+	cacheKeySyncPool.Put(k)
+}
+
+// GetHydratedTableDescriptor returns an ImmutableCopy with the types
+// hydrated. It may use a cached copy but all of the relevant type descriptors
+// will be retrieved via the TypeDescriptorResolver.
+func (c *Cache) GetHydratedTableDescriptor(
+	ctx context.Context, table *tabledesc.Immutable, res catalog.TypeDescriptorResolver,
+) (_ *tabledesc.Immutable, canUseCached bool, err error) {
+	if table.IsModified() {
+		return nil, false, nil
+	}
+	if !table.ContainsUserDefinedTypes() {
+		return table, true, nil
+	}
+
+	// TODO(ajwerner): This cache may thrash a bit right when a version of a type
+	// changes as different callers oscillate evicting the old and new versions of
+	// that type. It should converge rapidly but nevertheless, a finer-granularity
+	// cache could mitigate the problem. The idea would be to cache all tuples
+	// of table and type versions and then check if what we get from the resolver
+	// matches any of them.
+	var groupKey string // used as a proxy for cache hit
+	defer func() {
+		if err != nil || !canUseCached {
+			return
+		}
+		if groupKey == "" {
+			c.metrics.Hits.Inc(1)
+		} else {
+			c.metrics.Misses.Inc(1)
+		}
+	}()
+	key := newCacheKey(table)
+	defer func() {
+		if key != nil {
+			putCacheKey(key)
+		}
+	}()
+	for {
+		cached, ok := c.getHydratedTableDescriptorFromCache(ctx, key)
```

Comment:
I think that you only use `cached` in this scope, so maybe make it explicit `if cached, ok := ...`

Reply:
Done